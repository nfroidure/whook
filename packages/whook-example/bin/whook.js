#! /usr/bin/env node

import { join } from 'node:path';
import { URL } from 'node:url';
import { runCLI } from '@whook/whook';

const { prepareCommand } = await import(
  new URL(join('..', 'src', 'index.ts'), import.meta.url)
);

await runCLI(prepareCommand);
