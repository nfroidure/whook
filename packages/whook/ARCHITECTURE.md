[//]: # ( )
[//]: # (This file is automatically generated by the `jsarch`)
[//]: # (module. Do not change it elsewhere, changes would)
[//]: # (be overriden.)
[//]: # ( )
# Architecture Notes

## Summary

1. [Main file](#1-main-file)
   1. [Process run](#11-process-run)
   2. [Process preparation](#12-process-preparation)
   3. [Process environments](#13-process-environments)
2. [Services initializers](#2-services-initializers)
   1. [Base URL](#21-base-url)
   2. [IP detection](#22-ip-detection)
   3. [the `PWD` constant](#23-the-`pwd`-constant)
   3. [Port detection](#23-port-detection)
   4. [the `resolve` service](#24-the-`resolve`-service)
   5. [the `importer` service](#25-the-`importer`-service)
   6. [the `exit` service](#26-the-`exit`-service)
   6. [the `WHOOK_PLUGINS` constant](#26-the-`whook_plugins`-constant)
   7. [the `logger` service](#27-the-`logger`-service)
      1. [Logging](#271-logging)
   8. [The `CONFIG` service](#28-the-`config`-service)
   9. [the `$autoload` service](#29-the-`$autoload`-service)
      1. [the `APP_CONFIG` mapper](#291-the-`app_config`-mapper)
      2. [the `API` auto loading](#292-the-`api`-auto-loading)
         1. [Definitions loader](#2921-definitions-loader)
         2. [Typings generator](#2922-typings-generator)
         3. [Open API generator](#2923-open-api-generator)
      3. [the `INITIALIZER_PATH_MAP` mapper](#293-the-`initializer_path_map`-mapper)
         1. [Initializer path mapping](#2931-initializer-path-mapping)
      4. [the `ROUTES_HANDLERS` mapper](#294-the-`routes_handlers`-mapper)
      5. [the `ROUTES_WRAPPERS` auto loading](#295-the-`routes_wrappers`-auto-loading)
      6. [Service/handler/wrapper loading](#296-service/handler/wrapper-loading)
         1. [Plugins resolution](#2961-plugins-resolution)
         2. [Plugins/project paths](#2962-plugins/project-paths)
   10. [HTTP Transactions](#210-http-transactions)
      1. [New Transaction](#2101-new-transaction)
      2. [Transaction start](#2102-transaction-start)
      3. [Transaction errors](#2103-transaction-errors)
      4. [Transaction end](#2104-transaction-end)
   11. [HTTP Router](#211-http-router)
      1. [Coercion](#2111-coercion)
      2. [Validation](#2112-validation)
         1. [Schema validators](#21121-schema-validators)
      3. [Request body](#2113-request-body)
   12. [HTTP Server](#212-http-server)
   13. [Error handler](#213-error-handler)
      1. [Errors descriptors](#2131-errors-descriptors)
   14. [Local cron runner](#214-local-cron-runner)
3. [the routes](#3-the-routes)
   1. [Testing](#41-testing)


## 1. Main file

The Whook's main file exports :
- its specific types,
- its specific `knifecycle` compatible services,
- a few bootstrapping functions designed to be customizable.

[See in context](./src/index.ts#L165-L170)



### 1.1. Process run

Whook exposes a `runProcess` function to programmatically spawn
 its process. It is intended to be reusable and injectable so
 that projects can override the whole `whook` default behavior.

[See in context](./src/index.ts#L172-L176)



### 1.2. Process preparation

Whook exposes a `prepareProcess` function to create its
 configuration. It takes eventually additional injections that
 would be required at a higher level and a
 [Knifecycle](https://github.com/nfroidure/knifecycle)
 containing the bootstrapped environment and allowing
 to complete and run the process.

[See in context](./src/index.ts#L225-L232)



### 1.3. Process environments

The Whook `prepareEnvironment` function aims to provide the complete
 process environment without effectively planning its run. It allows
 to use that environment for testing or build purposes. It also
 provides a chance to override some services/constants
 before actually preparing the server in actual projects main file.

[See in context](./src/index.ts#L255-L261)



## 2. Services initializers

Whook's embed a few default initializers proxied from
 `common-services`, `@whook/*` or its own `src/services`
 folder. It can be wrapped or overridden, at will, later
 in a project using overrides.

[See in context](./src/index.ts#L272-L277)



### 2.1. Base URL

The `BASE_URL` service is intended to provide a base URL where
 the API can be found at. It can be overridden directly via
 injecting it but it is useful to have a usable URL while
 debugging production environnement.

[See in context](./src/services/BASE_URL.ts#L5-L10)



### 2.2. IP detection

If no `HOST` configuration is specified in dependencies nor in ENV,
 this service detects the machine host automagically.

[See in context](./src/services/HOST.ts#L7-L10)



### 2.3. the `PWD` constant

The Whook server heavily rely on the process working directory
 to dynamically load contents. We are making it available to
 the DI system as a constant.

[See in context](./src/index.ts#L303-L307)



### 2.3. Port detection

If no `PORT` configuration is specified in dependencies nor in ENV,
this service detects a free port automagically.

[See in context](./src/services/PORT.ts#L9-L12)



### 2.4. the `resolve` service

Whook uses the `common-services` `resolve` service to allow
 to easily mock/decorate all ESM resolutions.

[See in context](./src/index.ts#L311-L314)



### 2.5. the `importer` service

Whook uses the `common-services` `importer` service to allow
 to easily mock/decorate all ESM dynamic imports.

[See in context](./src/index.ts#L317-L320)



### 2.6. the `exit` service

Whook uses a built in `exit` service to allow
 to easily mock/decorate the app exit.

[See in context](./src/index.ts#L323-L326)



### 2.6. the `WHOOK_PLUGINS` constant

The Whook `$autoload` service needs to know where to look up
 for things like commands / routes /services etc...
The `WHOOK_PLUGINS` constant allows you to give the name of
 some modules that embrace the Whook folder structure allowing
 you to just install Whook's plugins to get them automatically
 loaded.

[See in context](./src/index.ts#L345-L352)



### 2.7. the `logger` service

Whook uses a built-in `logger` service to allow
 to easily route the application logs for the 
 `common-services` provided `log` service.

[See in context](./src/index.ts#L329-L333)



#### 2.7.1. Logging

Whook's default logger write to the NodeJS default console
   except for debugging messages where it uses the `debug`
   module so that you can set the `DEBUG` environment
   variable to `whook` and get debug messages in output:
  
  ```sh
  DEBUG=whook npx whook
  ```

[See in context](./src/services/logger.ts#L6-L15)



### 2.8. The `CONFIG` service

Loading the configuration files is done according to the `APP_ENV`
   environment variable. It basically requires a configuration hash
   where the keys are JSON formattable constants.

[See in context](./src/index.ts#L338-L342)



### 2.9. the `$autoload` service

Whook provides a simple way to load the constants, services
 and handlers of a project automatically though several
 strategies. It is done by implementing the `knifecycle`
 auto loading interface.

[See in context](./src/services/_autoload.ts#L73-L78)



#### 2.9.1. the `APP_CONFIG` mapper

First of all the autoloader looks for constants in the
 previously loaded `APP_CONFIG` configurations hash.

[See in context](./src/services/_autoload.ts#L213-L216)



#### 2.9.2. the `API` auto loading

We cannot inject the `API` in the auto loader since
 it is dynamically loaded so doing this during the auto
 loader initialization.

[See in context](./src/services/_autoload.ts#L126-L130)



##### 2.9.2.1. Definitions loader

The `DEFINITIONS` service provide a convenient way to
 gather your various definitions from the routes, crons
 or commands you created in the `src/(routes|cron|commands)`
 folder.

[See in context](./src/services/DEFINITIONS.ts#L23-L28)



##### 2.9.2.2. Typings generator

This command allows you to generate the API types that
 helps you to write your handler in a clean and safe
 manner.

[See in context](./src/commands/generateOpenAPITypes.ts#L14-L19)



##### 2.9.2.3. Open API generator

Here, we reuse the Open API handler to generate the
 definition of the API right inside a CLI command.

[See in context](./src/commands/generateOpenAPISchema.ts#L9-L13)



#### 2.9.3. the `INITIALIZER_PATH_MAP` mapper

The Whook auto-loader allows you to provide the file path
 of a service per its name. It exports a `WhookInitializerMap`
 type to help you ensure yours are valid.

[See in context](./src/services/_autoload.ts#L49-L54)



##### 2.9.3.1. Initializer path mapping

In order to be able to load a service from a given path map
 one can directly specify a path to use for its resolution.

[See in context](./src/services/_autoload.ts#L284-L287)



#### 2.9.4. the `ROUTES_HANDLERS` mapper

Here, we build the handlers map needed by the router by injecting every
 handler required by the API.

[See in context](./src/services/_autoload.ts#L231-L234)



#### 2.9.5. the `ROUTES_WRAPPERS` auto loading

We inject the `ROUTES_WRAPPERS_NAMES` in the `ROUTES_WRAPPERS`
 service so that they can be dynamically applied.

[See in context](./src/services/_autoload.ts#L253-L256)



#### 2.9.6. Service/handler/wrapper loading

Finally, we either load the handler/service/wrapper module
 if none of the previous strategies applied.

[See in context](./src/services/_autoload.ts#L278-L281)



##### 2.9.6.1. Plugins resolution

Whook auto loader can look for initializers in a list of
 plugins defined in the `WHOOK_PLUGINS` constant. This
 service checks their existence and computes once for all
 the details needed by the framework (to find services/handlers
 it provides).

[See in context](./src/services/WHOOK_RESOLVED_PLUGINS.ts#L7-L14)



##### 2.9.6.2. Plugins/project paths

Trying to load services from plugins/project paths.

[See in context](./src/services/_autoload.ts#L298-L300)



### 2.10. HTTP Transactions

[Whook](https://github.com/nfroidure/whook) takes a very
 unusual direction when it comes to dealing with HTTP
 transactions.
It makes requests and responses serializable (thanks to
 `WhookRequest` and `WhookResponse` types) to:

- only work with functions that take request and return
 responses (allowing your handlers to be pure functions),
- have  easily unit testable handlers thanks to concise
 snapshots.

This service is intended to build those literal objects 
 from Node HTTP ones (famously known as req/res) before
 passing them to the handlers. It also keeps track of
 running queries and ensure it is well handled by the
 server before releasing it. If not, the transaction is
 resolved with an error response (for timeouts or when
 an error were catched).

The `httpTransaction` service creates a new transaction
 for every single HTTP request incoming. It helps
 ensuring each request receives a response and avoid
 idle requests via a configurable timeout.

It is also a convenient abstraction of the actual
 request/response between the router and
 the NodeJS world. A common need is to fake the
 HTTP method for backward compatibility with old
 browsers/proxies by using the
 `X-HTTP-Method-Override` header.

You can simply do this by wrapping this service. See
 the @whook/method-override` module for a working
 example.

[See in context](./src/services/httpTransaction.ts#L58-L94)



#### 2.10.1. New Transaction

The idea is to maintain a hash of each pending
 transaction. To do so, we create a transaction
 object that contains useful informations about
 the transaction and we store it into the
 `TRANSACTIONS` hash.

Each transaction has a unique id that is either
 generated or picked up in the `Transaction-Id`
 request header. This allows to trace
 transactions end to end with that unique id.

[See in context](./src/services/httpTransaction.ts#L182-L193)



#### 2.10.2. Transaction start

Once initiated, the transaction can be started. It
   basically spawns a promise that will be resolved
   to the actual response or rejected if the timeout
   is reached.

[See in context](./src/services/httpTransaction.ts#L278-L283)



#### 2.10.3. Transaction errors

Here we are simply logging errors.
   It is important for debugging but also for
   ending the transaction properly if an error
   occurs.

[See in context](./src/services/httpTransaction.ts#L305-L310)



#### 2.10.4. Transaction end

We end the transaction by writing the final status
   and headers and piping the response body if any.

  The transaction can till error at that time but it
   is too late for changing the response status so
   we are just logging the event.
   This could be handled with
   [HTTP trailers](https://nodejs.org/api/http.html#http_response_addtrailers_headers)
   but the lack of client side support for now is
   preventing us to use them.

   Once terminated, the transaction is removed
from the `TRANSACTIONS` hash.

[See in context](./src/services/httpTransaction.ts#L354-L368)



### 2.11. HTTP Router

The Whook's `httpRouter` service  is responsible
 for wiring routes definitions to their actual
 implementation while filtering inputs and ensuring
 good outputs.

This is the default implementation of the Framework
 but it can be replaced or customized by setting your
 own configurations to replace the default ones
 (see the [API section](#API)).

The `httpRouter` service is responsible for handling
 the request, validating it and wiring the routes
 response to the actual HTTP response.

It is very opinionated and clearly diverges from the
 current standards based on a middlewares/plugins
 approach.

Here, the single source of truth is your API
 definition. No documentation, no route.

[See in context](./src/services/httpRouter.ts#L136-L158)



#### 2.11.1. Coercion

Data in headers / cookies / query string / path
 is defined as a string. So here, we are coercing
 it to their types. Objects and arrays of arrays
 ain't supported by design.

We do not go further in coercion since Whook's design
 is privileging strict APIs to avoid programming errors
 and only string coercion is required for the router.
 This is why we do not use AJV's coercion mechanism.

[See in context](./src/libs/coercion.ts#L3-L14)



#### 2.11.2. Validation

For performance reasons, the schemaValidators are
 created once for all at startup from the
 API definition.

One could argue that it would have been
 better for performances to generate
 the code statically. This is true. It
 may be done later but it won't change
 the way it works so, moving fast for
 now but keeping it in mind.

Also, looking closely to Prepack that
 could improve significantly this
 project performances with close to no
 time costs:
 https://github.com/facebook/prepack/issues/522#issuecomment-300706099

[See in context](./src/libs/validation.ts#L32-L50)



##### 2.11.2.1. Schema validators

Maintain a single place for JSON schema validation
 since it may repeat for several routes. Also
 warrantying that the same schema leads to the
 same reference for the Siso router parameters
 unicity checks.

[See in context](./src/services/schemaValidators.ts#L10-L17)



#### 2.11.3. Request body

According to the OpenAPI specification
there are two kinds of requests:
- **validated contents:** it implies to
 buffer their content and parse them to
 finally validate it. In that case, we
 provide it as a plain JS object to the
 routes handlers.
- **streamable contents:** often used
 for large files, those contents must
 be parsed and validated into the
 handler itself.

[See in context](./src/libs/body.ts#L19-L31)



### 2.12. HTTP Server

The Whook's `httpServer` service is responsible for
 instanciating the NodeJS HTTP Server and handling its
 start/shutdown.

It can be easily replaced by any other HTTP server
 (an HTTPS one for instance if you cannot use a
 gateway or a proxy to handle HTTPS connections).

The server takes in charge graceful shutdown by
 awaiting connections to be closed before shutting
 down which can take a long time (basically if a
 browser is still maintaining an open socket with
 it). You can short circuit this behavior, basically
 for development, by setting the `DESTROY_SOCKETS=1`
 environment variable.

[See in context](./src/services/httpServer.ts#L58-L75)



### 2.13. Error handler

Whook provides a default error handler that allows
 you to use the [`yerror`](https://github.com/nfroidure/yerror)
 and [`yhttperror`](https://github.com/nfroidure/yhttperror)
 modules to throw errors with some parameters allowing you
 to give some debugging context with raised errors.

Depending on the `NODE_ENV` it adds extra informations to
 HTTP response, beware to avoid activating it in production.

[See in context](./src/services/errorHandler.ts#L14-L24)



#### 2.13.1. Errors descriptors

Errors descriptors allows you to change the error
 handler behavior selectively for each error :
- change error codes to proper error messages templated
 with params,
- add URIs to help for debugging API errors
- override the error status with the `status` property.
- decide to pass some error parameters through with the
 `transmittedParams` so that your end users can have some
 more context on the error root cause.

[See in context](./src/services/errorHandler.ts#L26-L37)



### 2.14. Local cron runner

Whook allows you to run crons locally with the help
 of this local cron runner. Best practice is to
 export them and run it with some external runner
 like cloud ones, systemd or crontab.

[See in context](./src/services/localCronRunner.ts#L8-L14)



## 3. the routes

Routes are services that provide a definition and implements
 API routes.

[See in context](./src/routes/getOpenAPI.ts#L39-L42)



### 4.1. Testing

In such a hard life, Whook's make it simple to
 also test your commands.

[See in context](./src/commands/generateOpenAPISchema.test.ts#L6-L10)

