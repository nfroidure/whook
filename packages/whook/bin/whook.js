#! /usr/bin/env node

import path from 'path';
import runCLI from '../dist/cli.js';

const { prepareCommand } = await import(
  path.join(process.cwd(), 'dist', 'index.js')
);

await runCLI(prepareCommand);
