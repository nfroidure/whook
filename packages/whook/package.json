{
  "name": "@whook/whook",
  "version": "4.0.0-alpha.29",
  "description": "Build strong and efficient REST web services.",
  "metapak": {
    "configs": [
      "main",
      "readme",
      "eslint",
      "babel",
      "jest",
      "jsarch",
      "jsdocs",
      "typescript"
    ],
    "data": {
      "childPackage": true,
      "testsFiles": "'src/**/*.test.ts'",
      "files": "'src/**/*.ts'",
      "ignore": [
        "dist"
      ],
      "typesFiles": "src/**/*.ts",
      "typesDefs": "src/index.d.ts",
      "bundleFiles": [
        "dist/**/*.js",
        "src/index.d.ts",
        "dist",
        "src"
      ]
    }
  },
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "scripts": {
    "architecture": "jsarch 'src/**/*.ts' > ARCHITECTURE.md && git add ARCHITECTURE.md",
    "changelog": "conventional-changelog -p angular -i CHANGELOG.md -s",
    "cli": "env NODE_ENV=${NODE_ENV:-cli}",
    "compile": "babel --extensions '.ts,.js' src --out-dir=dist --source-maps=true",
    "cover": "npm run jest -- --coverage",
    "cz": "env NODE_ENV=${NODE_ENV:-cli} git cz",
    "doc": "echo \"# API\" > API.md; jsdoc2md undefined >> API.md && git add API.md",
    "jest": "NODE_ENV=test jest",
    "lint": "eslint 'src/**/*.ts'",
    "metapak": "metapak",
    "precz": "npm t && npm run lint && npm run metapak -- -s && npm run compile",
    "prettier": "prettier --write 'src/**/*.ts'",
    "preversion": "npm t && npm run lint && npm run compile && npm run metapak -- -s",
    "test": "npm run jest",
    "types": "rimraf -f 'dist/**/*.d.ts' && tsc --project . --declaration --emitDeclarationOnly --outDir dist"
  },
  "keywords": [
    "REST",
    "HTTP",
    "server",
    "web",
    "services",
    "OpenAPI"
  ],
  "author": {
    "name": "Nicolas Froidure",
    "email": "nicolas.froidure@insertafter.com",
    "url": "http://insertafter.com/en/index.html"
  },
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/nfroidure/whook.git"
  },
  "bugs": {
    "url": "https://github.com/nfroidure/whook/issues"
  },
  "homepage": "https://github.com/nfroidure/whook",
  "devDependencies": {
    "@babel/cli": "^7.8.3",
    "@babel/core": "^7.8.3",
    "@babel/plugin-proposal-class-properties": "^7.8.3",
    "@babel/plugin-proposal-object-rest-spread": "^7.8.3",
    "@babel/preset-env": "^7.8.3",
    "@babel/preset-typescript": "^7.8.3",
    "@babel/register": "^7.8.3",
    "@types/jest": "^24.9.0",
    "@typescript-eslint/eslint-plugin": "^2.16.0",
    "@typescript-eslint/parser": "^2.16.0",
    "babel-eslint": "^10.0.3",
    "babel-plugin-knifecycle": "^1.1.1",
    "eslint": "^6.8.0",
    "eslint-plugin-prettier": "^3.1.2",
    "jest": "^24.9.0",
    "jsarch": "^2.0.3",
    "jsdoc-to-markdown": "^5.0.3",
    "metapak": "^3.1.6",
    "metapak-nfroidure": "9.8.1",
    "prettier": "^1.19.1",
    "typescript": "^3.7.5"
  },
  "dependencies": {
    "@whook/http-router": "^4.0.0-alpha.29",
    "@whook/http-server": "^4.0.0-alpha.29",
    "@whook/http-transaction": "^4.0.0-alpha.29",
    "common-services": "^6.2.0",
    "debug": "^4.1.1",
    "dotenv": "^8.2.0",
    "internal-ip": "^5.0.0",
    "knifecycle": "^8.1.0",
    "openapi-types": "^1.3.5",
    "pkg-dir": "^4.2.0",
    "portfinder": "^1.0.25"
  },
  "contributors": [],
  "engines": {
    "node": ">=8.12.0"
  },
  "babel": {
    "presets": [
      "@babel/typescript",
      [
        "@babel/env",
        {
          "targets": {
            "node": "8.12.0"
          }
        }
      ]
    ],
    "plugins": [
      "@babel/proposal-class-properties",
      "@babel/plugin-proposal-object-rest-spread",
      "babel-plugin-knifecycle"
    ],
    "sourceMaps": true
  },
  "jest": {
    "coverageReporters": [
      "lcov",
      "html"
    ],
    "testPathIgnorePatterns": [
      "/node_modules/"
    ],
    "roots": [
      "<rootDir>/src"
    ]
  },
  "eslintConfig": {
    "extends": "eslint:recommended",
    "parserOptions": {
      "ecmaVersion": 2018,
      "sourceType": "module",
      "modules": true
    },
    "env": {
      "es6": true,
      "node": true,
      "jest": true,
      "mocha": true
    },
    "plugins": [
      "prettier"
    ],
    "rules": {
      "prettier/prettier": "error"
    },
    "parser": "@typescript-eslint/parser",
    "ignorePatterns": [
      "*.d.ts"
    ],
    "overrides": [
      {
        "files": [
          "*.ts"
        ],
        "rules": {
          "no-unused-vars": [
            1,
            {
              "args": "none"
            }
          ]
        }
      }
    ]
  },
  "prettier": {
    "semi": true,
    "printWidth": 80,
    "singleQuote": true,
    "trailingComma": "all",
    "proseWrap": "always"
  },
  "files": [
    "dist/**/*.js",
    "src/index.d.ts",
    "dist",
    "src",
    "LICENSE",
    "README.md",
    "CHANGELOG.md"
  ],
  "jsarch": {
    "parserOptions": {
      "plugins": [
        "typescript"
      ]
    }
  }
}
