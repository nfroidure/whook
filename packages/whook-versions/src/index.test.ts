/* eslint-disable @typescript-eslint/no-explicit-any */
import { jest, describe, it, expect, beforeEach } from '@jest/globals';
import {
  augmentAPIWithVersionsHeaders,
  initWrapHandlerWithVersionChecker,
} from './index.js';
import { initGetPing, initGetPingDefinition } from '@whook/whook';
import { YError } from 'yerror';
import { YHTTPError } from 'yhttperror';
import { NodeEnv } from 'application-services';
import type { AppEnvVars } from 'application-services';

const VERSIONS = [
  {
    header: 'X-API-Version',
    rule: '^1.0.0',
  },
  {
    header: 'X-SDK-Version',
    rule: '>=2.2.0',
  },
  {
    header: 'X-APP-Version',
    rule: '>=3.6.0',
  },
];

const ENV: AppEnvVars = { NODE_ENV: NodeEnv.Test };

describe('augmentAPIWithVersionsHeaders()', () => {
  it('should work', async () => {
    expect(
      await augmentAPIWithVersionsHeaders(
        {
          openapi: '3.1.0',
          info: {
            version: '1.0.0',
            title: 'Sample Swagger',
            description: 'A sample Swagger file for testing purpose.',
          },
          paths: {
            [initGetPingDefinition.path]: {
              [initGetPingDefinition.method]: initGetPingDefinition.operation,
            },
          },
        },
        VERSIONS,
      ),
    ).toMatchSnapshot();
  });
});

describe('wrapHandlerWithVersionChecker()', () => {
  const log = jest.fn();

  beforeEach(() => {
    log.mockReset();
  });

  it('should work with no version headers', async () => {
    const baseHandler = await initGetPing({
      ENV,
    });
    const wrapper = await initWrapHandlerWithVersionChecker({
      VERSIONS,
      log,
    });
    const wrappedHandler = await wrapper(baseHandler as any);
    const response = await wrappedHandler({});

    expect({
      response,
      logCalls: log.mock.calls,
    }).toMatchSnapshot();
  });

  it('should work with good api version headers', async () => {
    const baseHandler = await initGetPing({
      ENV,
    });
    const wrapper = await initWrapHandlerWithVersionChecker({
      VERSIONS,
      log,
    });
    const wrappedHandler = await wrapper(baseHandler as any);
    const response = await wrappedHandler({
      xApiVersion: '1.2.3',
    });

    expect({
      response,
      logCalls: log.mock.calls,
    }).toMatchSnapshot();
  });

  it('should work with good app version headers', async () => {
    const baseHandler = await initGetPing({
      ENV,
    });
    const wrapper = await initWrapHandlerWithVersionChecker({
      VERSIONS,
      log,
    });
    const wrappedHandler = await wrapper(baseHandler as any);
    const response = await wrappedHandler({
      xAppVersion: '3.6.0',
    });

    expect({
      response,
      logCalls: log.mock.calls,
    }).toMatchSnapshot();
  });

  it('should work with beta app version headers', async () => {
    const baseHandler = await initGetPing({
      ENV,
    });
    const wrapper = await initWrapHandlerWithVersionChecker({
      VERSIONS,
      log,
    });
    const wrappedHandler = await wrapper(baseHandler as any);
    const response = await wrappedHandler({
      xAppVersion: '4.0.0-beta.2',
    });

    expect({
      response,
      logCalls: log.mock.calls,
    }).toMatchSnapshot();
  });

  it('should work with good sdk version headers', async () => {
    const baseHandler = await initGetPing({
      ENV,
    });
    const wrapper = await initWrapHandlerWithVersionChecker({
      VERSIONS,
      log,
    });
    const wrappedHandler = await wrapper(baseHandler as any);
    const response = await wrappedHandler({
      xSdkVersion: '2.2.3',
    });

    expect({
      response,
      logCalls: log.mock.calls,
    }).toMatchSnapshot();
  });

  it('should fail with bad api version headers', async () => {
    const baseHandler = await initGetPing({
      ENV,
    });
    const wrapper = await initWrapHandlerWithVersionChecker({
      VERSIONS,
      log,
    });
    const wrappedHandler = await wrapper(baseHandler as any);

    try {
      await wrappedHandler({
        xApiVersion: '2.2.3',
      });
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect((err as YHTTPError).code).toEqual('E_DEPRECATED_VERSION');
      expect((err as YHTTPError).params).toEqual([
        'X-API-Version',
        '2.2.3',
        '^1.0.0',
      ]);
      expect((err as YHTTPError).httpCode).toEqual(418);
    }
  });

  it('should fail with bad app version headers', async () => {
    const baseHandler = await initGetPing({
      ENV,
    });
    const wrapper = await initWrapHandlerWithVersionChecker({
      VERSIONS,
      log,
    });
    const wrappedHandler = await wrapper(baseHandler as any);

    try {
      await wrappedHandler({
        xAppVersion: '0.0.0',
      });
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect((err as YHTTPError).code).toEqual('E_DEPRECATED_VERSION');
      expect((err as YHTTPError).params).toEqual([
        'X-APP-Version',
        '0.0.0',
        '>=3.6.0',
      ]);
      expect((err as YHTTPError).httpCode).toEqual(418);
    }
  });

  it('should fail with bad sdk version headers', async () => {
    const baseHandler = await initGetPing({
      ENV,
    });
    const wrapper = await initWrapHandlerWithVersionChecker({
      VERSIONS,
      log,
    });
    const wrappedHandler = await wrapper(baseHandler as any);

    try {
      await wrappedHandler({
        xSdkVersion: '0.2.3',
      });
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect((err as YHTTPError).code).toEqual('E_DEPRECATED_VERSION');
      expect((err as YHTTPError).params).toEqual([
        'X-SDK-Version',
        '0.2.3',
        '>=2.2.0',
      ]);
      expect((err as YHTTPError).httpCode).toEqual(418);
    }
  });
});
