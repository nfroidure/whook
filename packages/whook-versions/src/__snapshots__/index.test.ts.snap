// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`augmentAPIWithVersionsHeaders() should work 1`] = `
{
  "components": {
    "parameters": {
      "xApiVersion": {
        "example": "1.1.2-beta.1",
        "in": "header",
        "name": "X-API-Version",
        "required": false,
        "schema": {
          "pattern": "^(0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\.(0|[1-9]\\d*)(?:-((?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?$",
          "type": "string",
        },
      },
      "xAppVersion": {
        "example": "1.1.2-beta.1",
        "in": "header",
        "name": "X-APP-Version",
        "required": false,
        "schema": {
          "pattern": "^(0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\.(0|[1-9]\\d*)(?:-((?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?$",
          "type": "string",
        },
      },
      "xSdkVersion": {
        "example": "1.1.2-beta.1",
        "in": "header",
        "name": "X-SDK-Version",
        "required": false,
        "schema": {
          "pattern": "^(0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\.(0|[1-9]\\d*)(?:-((?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?$",
          "type": "string",
        },
      },
    },
  },
  "info": {
    "description": "A sample Swagger file for testing purpose.",
    "title": "Sample Swagger",
    "version": "1.0.0",
  },
  "openapi": "3.1.0",
  "paths": {
    "/ping": {
      "get": {
        "operationId": "getPing",
        "parameters": [
          {
            "$ref": "#/components/parameters/xApiVersion",
          },
          {
            "$ref": "#/components/parameters/xSdkVersion",
          },
          {
            "$ref": "#/components/parameters/xAppVersion",
          },
        ],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "pong": {
                      "enum": [
                        "pong",
                      ],
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Pong",
          },
        },
        "summary": "Checks API's availability.",
        "tags": [
          "system",
        ],
      },
    },
  },
}
`;

exports[`wrapRouteHandlerWithVersionChecker() should work with beta app version headers 1`] = `
{
  "logCalls": [
    [
      "debug",
      "游닌 - Initializing the version checker wrapper.",
    ],
  ],
  "response": {
    "body": {
      "pong": "pong",
    },
    "headers": {
      "X-Node-ENV": "test",
    },
    "status": 200,
  },
}
`;

exports[`wrapRouteHandlerWithVersionChecker() should work with good api version headers 1`] = `
{
  "logCalls": [
    [
      "debug",
      "游닌 - Initializing the version checker wrapper.",
    ],
  ],
  "response": {
    "body": {
      "pong": "pong",
    },
    "headers": {
      "X-Node-ENV": "test",
    },
    "status": 200,
  },
}
`;

exports[`wrapRouteHandlerWithVersionChecker() should work with good app version headers 1`] = `
{
  "logCalls": [
    [
      "debug",
      "游닌 - Initializing the version checker wrapper.",
    ],
  ],
  "response": {
    "body": {
      "pong": "pong",
    },
    "headers": {
      "X-Node-ENV": "test",
    },
    "status": 200,
  },
}
`;

exports[`wrapRouteHandlerWithVersionChecker() should work with good sdk version headers 1`] = `
{
  "logCalls": [
    [
      "debug",
      "游닌 - Initializing the version checker wrapper.",
    ],
  ],
  "response": {
    "body": {
      "pong": "pong",
    },
    "headers": {
      "X-Node-ENV": "test",
    },
    "status": 200,
  },
}
`;

exports[`wrapRouteHandlerWithVersionChecker() should work with no version headers 1`] = `
{
  "logCalls": [
    [
      "debug",
      "游닌 - Initializing the version checker wrapper.",
    ],
  ],
  "response": {
    "body": {
      "pong": "pong",
    },
    "headers": {
      "X-Node-ENV": "test",
    },
    "status": 200,
  },
}
`;
